core/commands/cid.go

```diff
@@ -0,0 +1,307 @@
+package commands
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"text/tabwriter"
+	"unicode"
+
+	cmds "gx/ipfs/QmPTfgFTo9PFr1PvPKyKoeMgBvYPh6cX3aDP7DHKVbnCbi/go-ipfs-cmds"
+	mhash "gx/ipfs/QmPnFwZ2JXKnXgMw8CdBPxn7FWh6LLdjUjxV1fKHuJnkr8/go-multihash"
+	cidutil "gx/ipfs/QmPyxJ2QS7L5FhGkNYkNcXHGjDhvGHueJ4auqAstFHYxy5/go-cidutil"
+	cmdkit "gx/ipfs/QmSP88ryZkHSRn1fnngAaV2Vcn63WUJzAavnRM9CVdU1Ky/go-ipfs-cmdkit"
+	mbase "gx/ipfs/QmSbvata2WqNkqGtZNg8MR3SKwnB8iQ7vTPJgWqB8bC5kR/go-multibase"
+	verifcid "gx/ipfs/QmVUhfewLZpSaAiBYCpw2krYMaiVmFuhr2iurQLuRoU6sD/go-verifcid"
+	cid "gx/ipfs/QmZFbDTY9jfSBms2MchvYM9oYRbAF19K7Pby47yDBfpPrb/go-cid"
+)
+
+var CidCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "Convert and discover properties of CIDs",
+	},
+	Subcommands: map[string]*cmds.Command{
+		"format": cidFmtCmd,
+		"base32": base32Cmd,
+		"bases":  basesCmd,
+		"codecs": codecsCmd,
+		"hashes": hashesCmd,
+	},
+}
+
+var cidFmtCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "Format and convert a CID in various useful ways.",
+		LongDescription: `
+Format and converts <cid>'s in various useful ways.
+
+The optional format string either "prefix" or or a printf style format string:
+` + cidutil.FormatRef,
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("cid", true, true, "Cids to format."),
+	},
+	Options: []cmdkit.Option{
+		cmdkit.StringOption("f", "Format string."),
+		cmdkit.StringOption("v", "CID version to convert to."),
+		cmdkit.StringOption("b", "Multibase to display CID in."),
+	},
+	Run: func(req *cmds.Request, resp cmds.ResponseEmitter, env cmds.Environment) {
+		fmtStr, _ := req.Options["f"].(string)
+		verStr, _ := req.Options["v"].(string)
+		baseStr, _ := req.Options["b"].(string)
+
+		opts := cidFormatOpts{}
+
+		switch fmtStr {
+		case "":
+			opts.fmtStr = "%s"
+		case "prefix":
+			opts.fmtStr = "%P"
+		default:
+			if strings.IndexByte(fmtStr, '%') == -1 {
+				resp.SetError(fmt.Errorf("invalid format string: %s", fmtStr), cmdkit.ErrNormal)
+				return
+			}
+			opts.fmtStr = fmtStr
+		}
+
+		switch verStr {
+		case "":
+			// noop
+		case "0":
+			opts.verConv = toCidV0
+		case "1":
+			opts.verConv = toCidV1
+		default:
+			resp.SetError(fmt.Errorf("invalid cid version: %s\n", verStr), cmdkit.ErrNormal)
+			return
+		}
+
+		if baseStr != "" {
+			encoder, err := mbase.EncoderByName(baseStr)
+			if err != nil {
+				resp.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+			opts.newBase = encoder.Encoding()
+		} else {
+			opts.newBase = mbase.Encoding(-1)
+		}
+
+		res, err := formatCids(req.Arguments, opts)
```

Comment:
I'd stream these results back. Users may want to feed quite a few CIDs into this command.

Reply:
I thought about that but didn't see the benefit.  This is a client side utility that will never block and is should be very fast, so the only benefit will be memory usage.  In addition the number of cid's passed in is limited by the length of command line so memory usage will never get out of control.

Reply:
Ah. Well, actually, arguments can be passed in on stdin (newline separated)... That's why I brought this up. Basically, a user may run something like `ipfs refs -u -r QmThing | ipfs cid format -f '%c' | sort -u`.

Reply:
I think you have to do something special to make that work.  Because when I tried it I get `Error: argument "cid" is required`.  I actually think this it is a bit of a misfeature to take the arguments from stdin without having to pass in a special flag, but that something to debate later.

Reply:
Fwiw, I agree it's a misfeature to take arguments from stdin without a flag -- I think it's pretty widespread convention now to use a single dash "`-`" argument to mean "read stdin instead of file".  (And that would make just as much sense for "read stdin instead of literal".)